rpc/namespaces/ethereum/eth/filters/filter_system.go

```diff
@@ -118,25 +115,27 @@ func (es *EventSystem) subscribe(sub *Subscription) (*Subscription, pubsub.Unsub
 	ctx, cancelFn := context.WithCancel(context.Background())
 	defer cancelFn()
 
-	existingSubs := es.eventBus.Topics()
-	for _, topic := range existingSubs {
-		if topic == sub.event {
-			eventCh, unsubFn, err := es.eventBus.Subscribe(sub.event)
-			if err != nil {
-				err := errors.Wrapf(err, "failed to subscribe to topic: %s", sub.event)
-				sub.err <- err
-				return nil, nil, err
-			}
+	// lock during the entire event system when subscribing.
+	es.indexMux.Lock()
+	defer es.indexMux.Unlock()
 
-			// wrap events in a go routine to prevent blocking
-			es.install <- sub
-			<-sub.installed
+	// try to subscribe to existing topic first
+	// no need to loop over the existing topics, just try to
+	// get it, an error is returned if the topic doesn't exist
+	eventCh, unsubFn, err := es.eventBus.Subscribe(sub.event)
+	if err == nil {
+		// topic exists, add subscription to the index
+		es.index[sub.typ][sub.id] = sub
+		sub.eventCh = eventCh
+		return sub, unsubFn, nil
+	}
 
-			sub.eventCh = eventCh
-			return sub, unsubFn, nil
-		}
+	if !errors.Is(err, pubsub.ErrTopicNotFound) {
+		sub.err <- err
+		return nil, nil, err
 	}
```

Comment:
If `err == nil` we will always fall into this condition. We need to do:
```
eventCh, unsubFn, err := es.eventBus.Subscribe(sub.event)
if err != nil {
    // There is an error. Check if ErrTopicNotFound.
    if !errors.Is(err, pubsub.ErrTopicNotFound) {
        // Error is NOT ErrTopicNotFound. Bubble up.
        sub.err <- err
		return nil, nil, err
    }    

    // We don't returned so error is ErrTopicNotFound.
    // Continue with execution and create topic.
} else {
    // No error. Topic exists so add subscription to the index.
    es.index[sub.typ][sub.id] = sub
    sub.eventCh = eventCh
	return sub, unsubFn, nil
}
```

Reply:
1e33bbc3